test/unit/test_environment.py

```diff
@@ -0,0 +1,160 @@
+    }}
+
+USER_HPS = dict(batch_size=32, learning_rate=.001)
+SAGEMAKER_HPS = {'sagemaker_region': 'us-west-2', 'default_user_module_name': 'net',
+                 'sagemaker_job_name': 'sagemaker-training-job', 'sagemaker_program': 'main.py',
+                 'sagemaker_submit_directory': 'imagenet', 'sagemaker_enable_cloudwatch_metrics': True,
+                 'sagemaker_container_log_level': logging.WARNING}
+
+ALL_HPS = dict(chain(USER_HPS.items(), SAGEMAKER_HPS.items()))
+
+
+@pytest.fixture(name='opt_ml_path')
+def override_opt_ml_path(tmpdir):
+    opt_ml = tmpdir.mkdir('opt').mkdir('ml')
+    with patch.dict('os.environ', {'BASE_PATH': str(opt_ml)}):
+        reload_module(environment)
+        yield opt_ml
+    reload_module(environment)
+
+
+@pytest.fixture(name='input_path')
+def override_input_path(opt_ml_path):
+    return opt_ml_path.mkdir('input')
+
+
+@pytest.fixture(name='input_config_path')
+def override_input_config_path(input_path):
+    return input_path.mkdir('config')
+
+
+@pytest.fixture(name='input_data_path')
+def override_input_data_path(input_path):
+    return input_path.mkdir('data')
+
+
+def test_read_json(tmpdir):
+    file_path = write_json('hyperparameters.json', ALL_HPS, tmpdir)
+
+    assert environment.read_json(file_path) == ALL_HPS
+
+
+def test_read_json_throws_exception():
+    with pytest.raises(IOError):
+        environment.read_json('non-existent.json')
+
+
+def test_read_hyperparameters(input_config_path):
+    write_json('hyperparameters.json', ALL_HPS, input_config_path)
+
+    assert environment.read_hyperparameters() == ALL_HPS
+
+
+def test_read_key_serialized_hyperparameters(input_config_path):
+    key_serialized_hps = {k: json.dumps(v) for k, v in ALL_HPS.items()}
+    write_json('hyperparameters.json', key_serialized_hps, input_config_path)
+
+    assert environment.read_hyperparameters() == ALL_HPS
+
+
+def test_split_hyperparameters_only_provided_by_user():
+    assert environment.split_hyperparameters(USER_HPS) == ({}, USER_HPS)
+
+
+def test_split_hyperparameters_only_provided_by_sagemaker():
+    assert environment.split_hyperparameters(SAGEMAKER_HPS) == (SAGEMAKER_HPS, {})
+
+
+def test_split_hyperparameters():
+    assert environment.split_hyperparameters(ALL_HPS) == (SAGEMAKER_HPS, USER_HPS)
+
+
+def test_resource_config(input_config_path):
+    write_json('resourceconfig.json', RESOURCE_CONFIG, input_config_path)
+
+    assert environment.read_resource_config() == RESOURCE_CONFIG
+
+
+def test_input_data_config(input_config_path):
+    write_json('inputdataconfig.json', INPUT_DATA_CONFIG, input_config_path)
+
+    assert environment.read_input_data_config() == INPUT_DATA_CONFIG
+
+
+def test_channel_input_dirs(input_data_path):
+    assert environment.channel_path('evaluation') == str(input_data_path.join('evaluation'))
+    assert environment.channel_path('training') == str(input_data_path.join('training'))
+
+
+@patch('subprocess.check_output', lambda s: b('GPU 0\nGPU 1'))
```

Comment:
https://sagemaker-build.corp.amazon.com/viewLog.html?buildId=5980&buildTypeId=ContainerSupport_Public_PullRequests&tab=buildLog&_focus=260#_state=156

Reply:
Your test dummy does not work because @patch requires either a second attribute or an extra argument to the decorated functions.

From the docs: https://github.com/testing-cabal/mock/blob/master/mock/mock.py#L1605

You can notice that all the testes are providing a lambda function that is the new object.

Tested using tox:
![screen shot 2018-04-19 at 7 33 18 am](https://user-images.githubusercontent.com/142074/38998225-f5f5e9ae-43a3-11e8-8c6b-9ceac5a73fca.png)

I am configuring teamcity today to have the same nice experience.

Reply:
i'm also surprised this works, but it does and makes waaay more sense than the usual patch usage!!